cli/internal/otel/receiver.go

```diff
@@ -0,0 +1,261 @@
+package otel
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.opentelemetry.io/collector/receiver"
+	"go.opentelemetry.io/collector/receiver/otlpreceiver"
+	noopmetric "go.opentelemetry.io/otel/metric/noop"
+	nooptrace "go.opentelemetry.io/otel/trace/noop"
+	"go.uber.org/zap"
+)
+
+type Span struct {
+	TraceID        string         `json:"trace_id"`
+	SpanID         string         `json:"span_id"`
+	TraceState     string         `json:"trace_state"`
+	ParentSpanID   string         `json:"parent_span_id"`
+	Name           string         `json:"name"`
+	StartTimestamp time.Time      `json:"start_timestamp"`
+	EndTimestamp   time.Time      `json:"end_timestamp"`
+	Attributes     map[string]any `json:"attributes"`
+}
+
+type Metric struct {
+	Name        string         `json:"name"`
+	Description string         `json:"description"`
+	Unit        string         `json:"unit"`
+	Value       int64          `json:"value"`
+	Attributes  map[string]any `json:"attributes"`
+}
+
+type ConsumeSpan = func(context.Context, Span)
+
+type ConsumeMetric = func(context.Context, Metric)
+
+type Consumer struct {
+	tracesFile    *os.File
+	consumeSpan   ConsumeSpan
+	metricsFile   *os.File
+	consumeMetric ConsumeMetric
+}
+
+func (c *Consumer) Shutdown(ctx context.Context) error {
+	var err error
+	if c.tracesFile != nil {
+		err = errors.Join(err, c.tracesFile.Close())
+	}
+	if c.metricsFile != nil {
+		err = errors.Join(err, c.metricsFile.Close())
+	}
+	return err
+}
+
+func newDefaultSpanConsumer(tracesFile *os.File) ConsumeSpan {
```

Comment:
I'm writing spans and metrics to files but we could do pretty much what we want we the data